### PR TITLE
a8n: Handle deletion of Changesets on codehosts

### PR DESCRIPTION
**Pull request by mrnugget**
First step towards #6396.

This changes the Syncer and the Bitbucket Server client to gracefully handle deleted changesets.

Since there's no way to tell from the Bitbucket Server response whether a changeset has been deleted or never created, we change the client to treat the "not found" it returns as a non-error.

In the Syncer we then check whether a given changeset has not been found and if so, we set its internal state to `DELETED`.

In the resolver for the `createChangesets` mutation we return an error.

There's a few different ways of going about this and I'm not really happy about returning the slice of "not found" changesets in `LoadChangesets`, but since it's a batch API, it's hard to do much better. Looking for suggestions.

What's missing:

- [x] Handling this correctly in the burndown chart. In order to do this, we need to record _when_ a changeset was deleted. I think adding a `deleted_at` column makes sense, since we then don't have to change internal state of the `Changeset.Metadata`.
- [x] UI presentation. I "fixed" it to make the linter happy, but right now a deleted changeset doesn't really look like it's been deleted and there's some weird UI glitches.

deleted_changeset

And also:

- Generating a timeline event for the deletion, as suggested in #6396. Since we don't have a timeline yet, I'm not sure whether we should already do that.

**Update on the last point**: With the `DeletedAt` timestamp I added it would be relatively easy to generate a `ChangesetEvent` in `(&Changeset).Events` that would then get persisted along with the other events. The caveat is, of course, that `DeletedAt` is the time that we _detect_ the deletion.

### CODE REVIEW

**Comment by tsenart** [comment]
Taking a look at this now.

**Comment by tsenart** [comment]
> Update on the last point: With the DeletedAt timestamp I added it would be relatively easy to generate a ChangesetEvent in (&Changeset).Events that would then get persisted along with the other events. The caveat is, of course, that DeletedAt is the time that we detect the deletion.

When we get integrated with the webhooks work @kzh did, we'll be able to know sooner.

File: cmd/repo-updater/repos/sources.go [hunk 1]
```diff
@@ -89,7 +89,10 @@ type Source interface {
 
 // A ChangesetSource can load the latest state of a list of Changesets.
 type ChangesetSource interface {
-	LoadChangesets(context.Context, ...*Changeset) error
+	// LoadChangesets loads the given Changesets from the sources and updates
+	// them. If a Changeset could not be found on the source, it's included in
+	// the returned slice.
+	LoadChangesets(context.Context, ...*Changeset) ([]*Changeset, error)
```
Comment (tsenart) [inline cmd/repo-updater/repos/sources.go]: I think we could avoid this interface change by returning a structured error that contains pointers to those `Changesets` that could not be found. Callers can then type switch on that error if and only if they need access to that information.

Something like:
```go
// LoadChangesetsError is an error type returned by LoadChangesets that
// helps discriminate what kind of error each of the passed Changesets
// produced, if any.
type LoadChangesetsError struct {
    NotFound []*Changeset
}

func (e LoadChangesetsError) Error() string { ... }
```

File: enterprise/internal/a8n/resolvers/resolver.go [hunk 2]
```diff
@@ -418,6 +419,21 @@ func (r *Resolver) CreateChangesets(ctx context.Context, args *graphqlbackend.Cr
 		return nil, err
 	}
 
+	errs := new(multierror.Error)
+	for _, c := range cs {
```
Comment (tsenart) [inline enterprise/internal/a8n/resolvers/resolver.go]: This is begging to be an utility method `func (cs Changesets) WithState(s State) Changesets` on `type Changesets []*Changeset`

**Comment by eseliger** [comment]
working on ui rn

**Comment by eseliger** [comment]
image

done reworking the states and getting rid of the collapsible for changesets without a diff. was there anything else on your mind to tick that box off?

**Comment by mrnugget** [comment]
> done reworking the states and getting rid of the collapsible for changesets without a diff. was there anything else on your mind to tick that box off?

Off the top of my head I'd say:

* make the colors more muted. Red and yellow do not signal "deleted" to me
* remove the `<a>` to the codehost since that will most likely lead to a 404

File: cmd/repo-updater/repos/sources.go (hunk 1 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/sources.go]: Nice. Thank you!

**Comment by eseliger** [comment]
> make the colors more muted. Red and yellow do not signal "deleted" to me

that's the `text-danger` we use globally in the web-app. Other options would be
- grey out the review icon
- grey out both, review and trash icon

opinions or other ideas?

**Comment by eseliger** [comment]
image

option with link disabled, greyed out review icon

image

option with both greyed out

**Comment by mrnugget** [comment]
I think the "both greyed out" option is fine.

**Comment by eseliger** [comment]
done

File: cmd/repo-updater/repos/sources.go (hunk 1 above)
Comment (mrnugget) [reply cmd/repo-updater/repos/sources.go]: See https://github.com/sourcegraph/sourcegraph/pull/7227/commits/8bbf4b2639dc76e317f3651bbfafe2a4c74b89e2

**Comment by mrnugget** [comment]
@christinaforney Since it's super rare that pull requests are deleted, I decided to merge this now instead of us tweaking the UI for longer. If you have any design suggestions, we can always follow-up.

**Comment by christinaforney** [comment]
@eseliger The red color is also used for closed changesets, right? If so, I think it makes sense to keep the greyed out state to avoid visual collision.

**Comment by eseliger** [comment]
yes and already done - thanks for confirming :)

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx [hunk 3]
```diff
@@ -38,89 +38,95 @@ export const ChangesetNode: React.FunctionComponent<ChangesetNodeProps> = ({
...
+                <ChangesetStateIcon
+                    className={`text-${changesetStatusColorClasses[changesetState]}`}
+                    data-tooltip={changesetStageLabels[changesetState]}
+                ></ChangesetStateIcon>
+            </div>
+            {node.__typename === 'ExternalChangeset' && (
+                <div className="flex-shrink-0 flex-grow-0 ml-1 mr-3">
+                    <ReviewStateIcon
+                        className={
+                            node.state === ChangesetState.DELETED
```
Comment (felixfbecker) [inline web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: I'd propose we include `DELETED` in the `changesetReviewStateColors` object (and maybe rename it) to save code here

**Pull request merged**